src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java

```diff
@@ -550,150 +550,100 @@ private static byte[] basesAfterFirst(final Allele altAllele) {
+                final int bufferPositionOfLastSnp = pdBytes.position() - 1;
+                final byte byteForLastSnp = pdBytes.get(bufferPositionOfLastSnp);
+                pdBytes.put(bufferPositionOfLastSnp,(byte) (byteForLastSnp | byteForThisSnp) );
+                continue;
+            } else if (event.getStart() == eventWithVariant.getStart() && useRef) {
+                // Ref alleles (even if they overlap undetermined events) should be skipped
                 continue;
             }
 
-            //Check that we are allowed to add this event (and if not we are)
-            if ((event.isIndel() && intermediateRefEndPosition - intermediateRefStartPosition < -1) || (!event.isIndel() && intermediateRefEndPosition - intermediateRefStartPosition < 0)) {//todo clean this up
-                throw new RuntimeException("Variant "+event+" is out of order in the PD event list: "+constituentEvents);
-            }
-
-            // Add the cigar for bases we skip over
-            if (intermediateRefEndPosition - intermediateRefStartPosition > 0) {
-                runningCigar.add(new CigarElement(intermediateRefEndPosition - intermediateRefStartPosition, CigarOperator.M));
-            }
-
-            // Include the ref base for indel if the base immediately proceeding this event is not already tracked
-            boolean includeRefBaseForIndel = event.isIndel() && (intermediateRefStartPosition <= intermediateRefEndPosition);
+            Utils.validate(basesBeforeNextEvent >= 0, () -> "PD event list: " + constituentEvents + " is out of order.");
 
-            // Determine the alleles to add
             Allele refAllele = event.refAllele();
             Allele altAllele = event.altAllele();
-            boolean isInsertion = altAllele.length() > refAllele.length(); // If its an insertion we flip to "ADD" the bases to the ref.
-            boolean isEvent = false;
-            byte[] basesToAdd;
-            // If this is the blessed variant, add
-            if (event.getStart()==eventWithVariant.getStart()) {
-                isEvent = true;
-                basesToAdd = useRef? refAllele.getBases() : altAllele.getBases();
-            // Otherwise make sure we are adding the longest allele (for indels) or the ref allele for snps.
-            } else {
-                basesToAdd = refAllele.length() >= altAllele.length() ? refAllele.getBases() : altAllele.getBases();
-            }
+            final int altRefLengthDiff = altAllele.length() - refAllele.length();
 
-            // Remove anchor base if necessary
-            if (event.isIndel() && !includeRefBaseForIndel) {
-                basesToAdd = Arrays.copyOfRange(basesToAdd, 1, basesToAdd.length);
-            }
+            boolean isInsertion = altRefLengthDiff > 0; // If its an insertion we flip to "ADD" the bases to the ref.
+            final boolean isEvent = event.getStart() == eventWithVariant.getStart();
```

Comment:
using evetn.getStart() seems dangerous (the old code is using it too i thought i wasn't) seems to preclude compound events where a snp and indel start on the same position but only one is determined? I would need to dig over the code again to see if its possible for those to happen but this equality assertion gives me some pause... I know compound UNDETERMINED events are allowed and perhaps thats why this isn't causing problems

Reply:
Now using `Event::equals`